state/state.go

```diff
@@ -127,12 +127,57 @@ func (s *State) Copy() *State {
 	}
 }
 
+var (
+	errNilGenesisDoc = errors.New("no genesisDoc was found")
+
+	genesisDBKey = []byte("genesis-doc")
+)
+
+// GenesisDoc is the accessor to retrieve the genesisDoc associated
+// with a state. If the state has no set GenesisDoc, it fetches from
+// its database the JSON marshaled bytes keyed by "genesis-doc", and
+// parses the GenesisDoc from that memoizing it for later use.
+// If you'd like to change the value of the GenesisDoc, invoke SetGenesisDoc.
+func (s *State) GenesisDoc() (*types.GenesisDoc, error) {
```

Comment:
Don't think that `GenesisDoc()` or `ChainID()` methods should return an error. If we know that genesis file is mandatory and TM will not work without it, maybe we can just throw a panic. That way we won't be needing to handle all these errors like https://github.com/tendermint/tendermint/pull/676/files#diff-80343e80650fe43a4394fc26afd5cb3bR508. I want to hear other opinions anyway.

Reply:
That's a justified concern, it makes sense to simplify the PR more, which is nice. However, the requirements of the PR request that we instead have a way of alternatively loading the genesisDoc and chainID after loading them from a DB is what warrants us returning an error
since:
1) Non-existent data
2) Invalid data stored

Otherwise I'd have agreed that the PR can just be simplified to a single set and panic as you advise.

Reply:
Agree, we should not return an error. This is a panic scenario.